narrative_llm_agent/user_interface/ui_dash_hitl.py

```diff
@@ -612,15 +684,67 @@ def run_analysis_planning_callback(proceed_clicks, credentials, collected_metada
             "Missing narrative ID or reads ID. Please complete metadata collection or manual input.",
             color="warning",
         ), True
-
-    # Run the analysis planning
-    with StreamRedirector(ANALYSIS_LOG_BUFFERS[session_id]):
-        print("Starting KBase workflow planning")
-        # with open(os.path.dirname(os.path.abspath(__file__)) + "/temp.json") as in_json:
-        #     result = json.load(in_json)
-        result = run_analysis_planning(narrative_id, reads_id, description, credentials)
-
+    # Setup streaming based on environment
+    if redis_client:
+        # Redis-based streaming for production
+        stream_redirector = RedisStreamRedirector(session_id, redis_client)
+        
+        # Log initial message to Redis
+        stream_redirector.write("🚀 Starting KBase workflow planning...\n")
+        stream_redirector.write(f"📋 Session ID: {session_id}\n")
+        stream_redirector.write(f"🔬 Narrative ID: {narrative_id}\n")
+        stream_redirector.write(f"📊 Reads ID: {reads_id}\n")
+        stream_redirector.write("=" * 50 + "\n")
+        
+    else:
+        # Local buffer for development
+        if session_id not in ANALYSIS_LOG_BUFFERS:
+            ANALYSIS_LOG_BUFFERS[session_id] = StringIO()
+        stream_redirector = StreamRedirector(ANALYSIS_LOG_BUFFERS[session_id])
+        
+        print("🚀 Starting KBase workflow planning")
+        print(f"📋 Session ID: {session_id}")
+        print(f"🔬 Narrative ID: {narrative_id}")
+        print(f"📊 Reads ID: {reads_id}")
+        print("=" * 50)
+        # Run the analysis planning
+        with StreamRedirector(ANALYSIS_LOG_BUFFERS[session_id]):
+            print("Starting KBase workflow planning")
+            # with open(os.path.dirname(os.path.abspath(__file__)) + "/temp.json") as in_json:
+            #     result = json.load(in_json)
+            result = run_analysis_planning(narrative_id, reads_id, description, credentials)
     del ANALYSIS_LOG_BUFFERS[session_id]
+    try:
+        # Redirect stdout to our streaming mechanism
+        import sys
+        original_stdout = sys.stdout
+        sys.stdout = stream_redirector
+        
+        try:
+            result = run_analysis_planning(narrative_id, reads_id, description, credentials)
+        finally:
+            sys.stdout = original_stdout
+            
+        # Log completion
+        if redis_client:
+            stream_redirector.write(f"\n✅ Analysis planning completed with status: {result.get('status', 'unknown')}\n")
+        else:
+            print(f"\n✅ Analysis planning completed with status: {result.get('status', 'unknown')}")
```

Comment:
The flow above looks a little confusing to me.
You could probably simplify it with the following:

Make the RedisStreamRedirector act as a context manager (with a __enter__ and __exit__ functions).
Set a stream_redirector based on conditions.
Run it the same way with either.
Something like:

```
if using_redis:
    stream_redirector = RedisStreamRedirector(session_id, redis_client)
elif (some other? diskcache?):
    stream_redirector = OtherStreamRedirector()
else:
    stream_redirector = StreamRedirector()

with stream_redirector:
    print(intro stuff)
    result = run_analysis_planning(...)
````

And the rest should just clean itself up.

Reply:
Same goes in the analysis_approval module.